### ISSUE TITLE
Publish binaries to NPM

### ISSUE THREAD

**Issue by roninjin10** [foundry-rs/foundry#4337]
### Component

Forge

### Describe the feature you would like

I would like to be able to download forge as a binary npm package for easier integration into Frontend and Node applications.  I am unclear on if it is even possible to package up forge as a standalone binary   

### Additional context

I would like to be able to download forge as a binary npm package for easier integration into Frontend and Node applications.   As an example, if forge is a part of a build pipeline it becomes very awkward to integrate it into a vercel build.  

I am happy to contribute to making this happen but I would need some guidance on
1. Is it even possible to package up forge into a binary?   Would I have to try to package up foundry-up and then execute foundry-up?
2. A pointer to where to look in foundry to get started

My specific use case is Iam working to build a forge inspired ethers.js alternative.  User would be able to write forge scripts.  The API is similar to this

```typescript
import { mutate } from 'ts-sol'
import { TransferEntireBalance } from './ERC20.s.sol'

....

<button onClick={() => mutate(TransferEntireBalance, { signer })} />
```

I want to reuse forge as much as possible but if I can't find a way to get forge on NPM I'm likely going to be forced to reimplement cheat codes into a javascript based VM or hardhat.

**Comment by mattsse** [comment]
there's https://github.com/foundry-rs/hardhat/tree/develop/packages/hardhat-forge 

and https://github.com/foundry-rs/hardhat/tree/develop/packages/easy-foundryup

**Comment by roninjin10** [comment]
Interesting I don't think hardhat-forge is super useful.  We use that at OP Labs and I believe foundry is still a required dependency for it to work.

Easy foundry-up looks promising though.  I will dig into that more later

**Comment by roninjin10** [comment]
@tynes any thoughts on this issue?

**Comment by roninjin10** [comment]
Tried using easy-foundryup in vercel and looks like there are some issues getting foundry to do it's thing on vercel likely because of missing deps in the container
 
<img width="1223" alt="image" src="https://user-images.githubusercontent.com/35039927/218325936-ec3c37c2-87dc-483d-98d3-c4a7fdc5adec.png">
```
foundryup: done
--
09:09:43.978 | forge: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by forge)
09:09:43.978 | forge: /lib64/libc.so.6: version `GLIBC_2.29' not found (required by forge)
09:09:43.979 | forge: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by forge)
09:09:43.979 | forge: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by forge)
09:09:43.981 | ELIFECYCLE  Command failed with exit code 1.
09:09:44.000 | Error: Command "pnpm run build" exited with 1
```

**Comment by roninjin10** [comment]
I think the more general question here is is it possible to package up forge into a binary such that it doesn't require external dependencies to execute?

**Comment by onbjerg** [comment]
Vercel is likely using musl. You either need to install glibc or follow these instructions: https://github.com/foundry-rs/foundry#out-of-date-glibc-error-when-running-forge-from-default-foundryup-install

You will not be able to use Forge as you described in your issue though, as Forge is not distributable on npm and is not importable as a module in JavaScript, as we do not support wasm.

**Comment by roninjin10** [comment]
@onbjerg you don’t need to be javascript or wasm to be on npm.   Any binary can be on npm

**Comment by roninjin10** [comment]
Also I’m trying to use forge as a build tool.    Using it in the browser, that isn’t what I’m trying to do (yet).    I am trying to build a way of using forge like cheat codes in the browser but I’m building that myself in top of Ethereum js.   Forge is for building contracts at buildtime not runtime though

**Comment by roninjin10** [comment]
Thanks helpful to know the timeline.   What I think I will do is use that toolchain to create a NODE18 docker image that anybody can build in for now.   It's not perfect, but hopefully works for most folks

**Comment by o-az** [comment]
FYI you don’t need to bundle anything. You can publish the repository exactly as it is to npm and Github Packages.

**Comment by roninjin10** [comment]
Wouldn’t that mean you had to build it though?   It takes a very long time to build

**Comment by o-az** [comment]
Yes, I was commenting on the fact you that bundling is not a requirement to publish to npm or similar registries.

I think having an action that publishes the precompiled binaries which are available in the release page would be nice.

**Comment by alex-grover** [comment]
I think it's worth reopening this issue, or otherwise finding a way to enable the following workflow:

I'm using the [Wagmi CLI foundry plugin](https://wagmi.sh/cli/plugins/foundry) to generate app code from my foundry project. I want to do this at build time on Vercel, so that configuration (like deployed contract addresses) is always up to date.

I imagine this is one of the most common dev setups right now and it doesn't seem like it's possible.

**Comment by vmaark** [comment]
> I think it's worth reopening this issue, or otherwise finding a way to enable the following workflow:
> 
> I'm using the [Wagmi CLI foundry plugin](https://wagmi.sh/cli/plugins/foundry) to generate app code from my foundry project. I want to do this at build time on Vercel, so that configuration (like deployed contract addresses) is always up to date.
> 
> I imagine this is one of the most common dev setups right now and it doesn't seem like it's possible.

Indeed, I have the same issue. The only alternative now seems to be committing the generated files which is suboptimal.

**Comment by roninjin10** [comment]
I agree this issue is still worth reopening. NPM is for distributing anything not just javascript

**Comment by roninjin10** [comment]
@onbjerg any chance we could reopen this issue? I’m happy to help

**Comment by sambacha** [comment]
we will not be silenced 

also this goes for @forge-std package tooo

**Comment by zerosnacks** [comment]
> also this goes for @forge-std package

Good point, cross-linking https://github.com/foundry-rs/forge-std/issues/170

@o-az and I will be looking into this for the next Foundry release

**Comment by grandizzy** [comment]
I am looking into this and I *think* the best way to do it is as explained in this blog post https://sentry.engineering/blog/publishing-binaries-on-npm#exploring-our-options
@o-az and others in discussion, would this make sense? Thanks!

**Comment by zerosnacks** [comment]
Considering the binaries are indeed quite large, totalling > 250mb zipped for all it is indeed a question whether it makes sense to bundle all versions into one package rather than install dynamically. A binary package I worked on before did not have that issue which is why you could just pick the correct one at runtime.

There are security considerations given that `postinstall` runs in user space without any sandboxing. Ensuring the integrity of the downloaded binary is very important here. Users can also disable postinstall scripts from running using the `--ignore-scripts` flag.

https://www.nodejs-security.com/blog/npm-ignore-scripts-best-practices-as-security-mitigation-for-malicious-packages 

@o-az would this work for you?

**Comment by roninjin10** [comment]
IMO the best practices here can been seen in esbuild and is pretty much exactly your plan

- Esbuild publishes [1 npm package for every architecture](https://github.com/evanw/esbuild/tree/main/npm/%40esbuild)
- Then they publish [the main package](https://github.com/evanw/esbuild/tree/main/npm/esbuild)
- That main package [has a postinstall script](https://github.com/evanw/esbuild/blob/main/npm/esbuild/package.json#L10)
- That package then will dynamically [install from npm based on architecture](https://github.com/evanw/esbuild/blob/main/lib/npm/node-install.ts)

Note: however that esbuild is doing a lot of work here that you might want to deprioritize. 

Now myself as an end user I can npm install the main package for convenience. Or instead I can install directly the package that fits my architecture if I'd rather have control over that process and not run postinstall scripts.

**Comment by roninjin10** [comment]
I want to share as a follow up, [napi-rs](https://github.com/napi-rs/napi-rs) is a way to not only get your package onto NPM but also make it usable within node.js. I would call this nice to have but figured I'd share it as [tevm](https://github.com/evmts/tevm-monorepo/) will be using it as we start adding rust packages.

**Comment by zerosnacks** [comment]
From a conversation with @pcaversaccio, we must make sure to generate a provenance statement in our release flow; this is similar to our attestation artifacts for our regular binaries. I am not sure if we already do this in the latest version of the proposed workflow.

https://docs.npmjs.com/generating-provenance-statements

https://github.com/pcaversaccio/snekmate/blob/main/.github%2Fworkflows%2Fpublish-npm.yml

It looks like this on NPM: https://www.npmjs.com/package/snekmate#provenance

Reference: https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5644

cc @grandizzy / @o-az

**Comment by zerosnacks** [comment]
We've decided to deprecate the NPM publishing pipeline for the following reasons:

- **Lack of traction**: npm has not become a meaningful distribution channel for Foundry.
- **Low download numbers**: Adoption of the `@foundry-rs/*` packages is minimal, and `foundryup` remains the canonical install path for the vast majority of users.
- **Pipeline complexity**: The publishing workflow (per-arch packages, meta packages, provenance/auth, staging from release artifacts) carries significant maintenance overhead.
- **Maintenance distraction**: The npm publishing surface continually attracts a stream of automated/AI-generated PRs proposing small drive-by changes, which distract maintainers from the project's core focus.

What "deprecation" means here:

This only stops publishing **new** versions. Existing `@foundry-rs/*` packages already on npm will remain available — users who depend on previously published versions are not affected. No packages will be unpublished or removed from the registry.
